### ISSUE TITLE
Possible Misalignment Between Critic Model's Value Calculation and Attention Mask in PPO Training?

### ISSUE THREAD

**Issue by tiesanguaixia** [alibaba/ROLL#3]
Thank you for the wonderful work! 

I'm investigating the value calculation in the PPO training process within the ROLL framework and have a question regarding the alignment between critic model outputs and attention masks.

From the critic model's loss calculation perspective:
1. The values from the forward pass are taken excluding the first element  [`values[:,1:]`](https://github.com/alibaba/ROLL/blob/70800b83f416fc98728654f12258425a359a033a/roll/pipeline/base_worker.py#L485C8-L486C1) 
2. These values are then processed with returns and combined with the [`response_mask`](https://github.com/alibaba/ROLL/blob/70800b83f416fc98728654f12258425a359a033a/roll/pipeline/base_worker.py#L481) via masked mean

After analyzing the code, I noticed what appears to be a potential misalignment in the indexing:
- For a response containing 4 valid tokens (including `<|im_end|>`), I believe we should be calculating values from the last prompt token up to (but not including) the `<|im_end|>` token
- However, the current implementation seems to be calculating values from the first response token through to the `<|im_end|>` token

Could someone clarify if:
1. My understanding of the expected behavior is correct?
2. There's a potential indexing issue here, or if I'm misunderstanding the implementation?

This appears in the PPO training pipeline's value calculation and loss computation. Any insights would be appreciated!

**Comment by PanAndy** [comment]
Let me think. 
Please wait🤔

**Comment by PanAndy** [comment]
<img width="831" alt="Image" src="https://github.com/user-attachments/assets/2db34258-ea4c-4413-81cc-cabeef07d11c" />

it seem like there's no issue here?

**Comment by tiesanguaixia** [comment]
> <img alt="Image" width="831" src="https://private-user-images.githubusercontent.com/15311784/449478992-2db34258-ea4c-4413-81cc-cabeef07d11c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg2ODY4NTgsIm5iZiI6MTc0ODY4NjU1OCwicGF0aCI6Ii8xNTMxMTc4NC80NDk0Nzg5OTItMmRiMzQyNTgtZWE0Yy00NDEzLTgxY2MtY2FiZWVmMDdkMTFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTMxVDEwMTU1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE0NTUzNjI1NzJmOTkxYzg5N2E1YmExZWViOWE2OGZmNDFlNDAzOTg1Yzk3YzA2MmE0OTJiZTU3MjRmZDUxMzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.o2ed8Lx4RzQJQz8-JYqwtT2QKK1aJcdUFc55gJxJPTc">
> it seem like there's no issue here?

Thank you for the explanation! I'm still slightly confused and have created a diagram to illustrate my understanding. Could you please check if I've misunderstood anything?

<img width="1246" alt="Image" src="https://github.com/user-attachments/assets/e2227252-88bc-4822-811d-31824e4cb578" />

**Comment by wwxFromTju** [comment]
Hey, first of all, thank you for your interest in ROLL.

Regarding your question, I am currently confused as to why we need to consider the "v5" corresponding to "p5". Here "p5" is not the token generated by the model. In fact, we should optimize the tokens generated by the model. 

Therefore, what we really care about is r1 until im_end.

**Comment by tiesanguaixia** [comment]
> Hey, first of all, thank you for your interest in ROLL.
> 
> Regarding your question, I am currently confused as to why we need to consider the "v5" corresponding to "p5". Here "p5" is not the token generated by the model. In fact, we should optimize the tokens generated by the model.
> 
> Therefore, what we really care about is r1 until im_end.

From my perspective, `V5` is calculated from the logits corresponding to input `P5`, and the logits corresponding to input `P5` are `R1` derived from. For reference, I've consulted this [technical explanation on Zhihu](https://zhuanlan.zhihu.com/p/14569025663) and OpenRLHF's implementation:

https://github.com/OpenRLHF/OpenRLHF/blob/494850f50342ed38d5ae76ef45a3207f3523b582/openrlhf/models/model.py#L280-L285

**Comment by wwxFromTju** [comment]
I guess you may have confused the structure of actor and critic.

The actor forwards logits, and then we select the next token. Then we need to note that the critic model can forward a scale value for each token.

So, for the critic model, we need to consider the actual selected token and then forward its corresponding value.

**Comment by wwxFromTju** [comment]
@tiesanguaixia  Oh, I see. What you want to get is $V(s_{0})$. I misunderstood your intention at first. From the perspective of taking $V(s_{0})$, we need to take the value corresponding to p5. 

Please check this place @PanAndy.

**Comment by PanAndy** [comment]
@wwxFromTju okk。
Thank you @tiesanguaixia  for pointing that out. 🫡
I just cross-checked with the internal version and there's indeed a discrepancy here. Let me make an adjustment.
